api/include/opentelemetry/baggage/baggage.h

```diff
@@ -0,0 +1,66 @@
+#include "opentelemetry/common/kv_properties.h"
+#include "opentelemetry/nostd/shared_ptr.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace baggage
+{
+
+class Baggage
+{
+
+public:
+  Baggage() : kv_properties_(new opentelemetry::common::KeyValueProperties()) {}
+
+  template <class T>
+  Baggage(const T &keys_and_values)
+      : kv_properties_(new opentelemetry::common::KeyValueProperties(T))
+  {}
+
+  bool GetValue(const nostd::string_view &key, std::string &value)
+  {
+    return kv_properties_->GetValue(key, value);
+  }
+
+  // @return all key-values entris by repeatedly invoking the function reference passed as argument
+  // for each entry
+  bool GetAllEntries(
+      nostd::function_ref<bool(nostd::string_view, nostd::string_view)> callback) const noexcept
+  {
+    return kv_properties_->GetAllEntries(
+        callback);  // [&callback](nostd::string_view key, nostd::string_view value) {
+  }
+
+  nostd::shared_ptr<Baggage> Delete(const nostd::string_view &key)
+  {
+    nostd::shared_ptr<Baggage> baggage(new Baggage(kv_properties_->Size() - 1));
+    kv_properties_->GetAllEntries(
+        [&baggage, &key](nostd::string_view e_key, nostd::string_view e_value) {
+          if (key != e_key)
+            baggage->kv_properties_->AddEntry(e_key, e_value);
+          return true;
+        });
+    return baggage;
+  }
+
+  nostd::shared_ptr<Baggage> Set(const nostd::string_view &key, const nostd::string_view &value)
+  {
+    nostd::shared_ptr<Baggage> baggage(new Baggage(kv_properties_->Size() + 1));
+    baggage->kv_properties_->AddEntry(key, value);
```

Comment:
I have a question here about the key_value pairs in the baggage. With this implementation, we are adding duplicate keys in our baggage (eg. if key already existed in the baggage, we still prepend a new entry with the same key). Do we want to keep this behaviour or we should avoid adding old entry in our returned baggage.
Standard says to keep only one value for each baggage name ref:  https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md#overview
We are always returning fresh value in getter so in a way it does not seem to be a big issue. If we keep this behaviour same, I think we need to change couple of things in Delete function as size could be less than kv_properties_->Size() - 1 (line 37)

Let me know what do you think about this. This is also important because I was thinking if we could keep baggage entries as linked list instead of array to save space. Please correct me if I am missing anything.

Reply:
Good point, we should first add the k-v pair from current baggage except for key passed as argument ( if already exist). And in the end, add the new k-v. It should be fine to store one less element than actual allocation as long as we track number of elements internally.

Reply:
Yes, I think that should be the way. One more thing, I am planning to use linked list (deep clone everytime) in kv_properties for optimum memory usage (as I don't see usecase of any random access). Currently, context also uses linked list but its not deep clone as context does not have delete key api. So, I context code will be unchanged in this refactoring.
Do you agree with the approach ? Wanted to give you heads up before making change to make sure that I am not missing anything.

Reply:
What benefit do you see with linked list as compared to fixed size array ( which we use now )? Is it because of extra processing we may have to do initially to find exact allocation size in case of array? We want to minimize multiple dynamic memory allocation in api. And iterating a linked list would be slow as compared to array due to non-contiguous memory allocation of it's node.

Reply:
@pyohannes  - would like to hear your thoughts here.

Reply:
Yeah, I agree that linked list solution will be slow. I think, the main concern behind it is when we parse headers to extract the baggage, we do not know the size of key value pairs. So, solution is that either we parse header two times or maintain dynamic size data structure. I see that, in trace_state, you have made two passes.
We might hit invalid entries in between and so actual entries in baggage might be lesser. So, essentially first pass gives upper bound on number of entries.

Reply:
Nikhil- That's true. just wondering if you have any further questions on this design ?

Reply:
Should I keep the array structure only and not the linked list ?

Reply:
I think, I can also clarify thing about baggage metadata here. We have not incorporated baggage metadata here. In my solution, I had kept baggage metadata along with key, value. Its an opaque structure for now so languages like golang are not keeping it separately (keeping it as a part of value). How should we proceed, I am kind of stuck there. We might want to make kv_properties more generic if we decide to keep it separately.

Reply:
Yes that's correct.

Reply:
> I think, I can also clarify thing about baggage metadata. We have not incorporated baggage metadata here. In my solution, I had kept baggage metadata along with key, value. Its an opaque structure with no semantic meaning for now so languages like golang are not keeping it separately (keeping it as a part of value). How should we proceed ? I am kind of stuck there. We might want to make kv_properties more generic if we decide to keep it separately.

I won't be much worried about incorporating metadata as of now, and probably revisit once it's semantics/usage is properly defined by W3C. Don't see languages like python, dotnet supporting it too.